File: frontend/src/services/masterStats.ts
import api from './api';

// Types
export interface DailyGrowth {
  date: string;
  learning_sessions: number;
  grammar_sessions: number;
  new_students: number;
}

export interface SystemOverview {
  total_students: number;
  total_teachers: number;
  total_learning_sessions: number;
  total_grammar_sessions: number;
  total_learning_answers: number;
  total_grammar_answers: number;
  total_words: number;
  total_grammar_questions: number;
  daily_growth: DailyGrowth[];
}

export interface WordCalibrationItem {
  word_id: string;
  english: string;
  korean: string;
  book_name: string;
  lesson: string;
  curriculum_level: number;
  actual_accuracy: number;
  attempt_count: number;
  avg_time_sec: number | null;
  suggested_level: number;
  gap: number;
}

export interface GrammarCalibrationItem {
  question_id: string;
  question_type: string;
  question_type_label: string;
  book_title: string;
  chapter_title: string;
  assigned_difficulty: number;
  actual_accuracy: number;
  attempt_count: number;
  avg_time_sec: number | null;
  suggested_difficulty: number;
  gap: number;
}

export interface CalibrationResponse {
  word_calibrations: WordCalibrationItem[];
  grammar_calibrations: GrammarCalibrationItem[];
}

export interface BadQuestionItem {
  question_id: string;
  question_type: string;
  question_type_label: string;
  book_title: string;
  chapter_title: string;
  difficulty: number;
  accuracy: number;
  attempt_count: number;
  avg_time_sec: number | null;
  flag_reason: string;
}

export interface BadWordItem {
  word_id: string;
  english: string;
  korean: string;
  book_name: string;
  curriculum_level: number;
  accuracy: number;
  attempt_count: number;
  flag_reason: string;
}

export interface BadQuestionResponse {
  grammar_issues: BadQuestionItem[];
  word_issues: BadWordItem[];
}

export interface ConfusedPair {
  correct_answer: string;
  wrong_answer: string;
  confusion_count: number;
}

export interface QuestionTypeAccuracy {
  question_type: string;
  label: string;
  total: number;
  correct: number;
  accuracy_pct: number;
  avg_time_sec: number | null;
}

export interface ErrorPatternResponse {
  confused_word_pairs: ConfusedPair[];
  word_question_type_breakdown: QuestionTypeAccuracy[];
  grammar_question_type_breakdown: QuestionTypeAccuracy[];
}

export interface StageCount {
  stage: number;
  count: number;
}

export interface StageDuration {
  stage: number;
  avg_days: number;
}

export interface SrsOptimizationData {
  stage_distribution: StageCount[];
  avg_days_per_stage: StageDuration[];
  total_mastered: number;
}

export const masterStatsService = {
  getOverview: () => api.get<SystemOverview>('/api/v1/master-stats/overview').then(r => r.data),
  getCalibration: () => api.get<CalibrationResponse>('/api/v1/master-stats/calibration').then(r => r.data),
  getBadQuestions: () => api.get<BadQuestionResponse>('/api/v1/master-stats/bad-questions').then(r => r.data),
  getErrorPatterns: () => api.get<ErrorPatternResponse>('/api/v1/master-stats/error-patterns').then(r => r.data),
  getSrsData: () => api.get<SrsOptimizationData>('/api/v1/master-stats/srs-data').then(r => r.data),
};

File: backend/app/schemas/master_stats.py
"""Master statistics schemas."""
from typing import Optional
from pydantic import BaseModel


class DailyGrowth(BaseModel):
    date: str
    learning_sessions: int
    grammar_sessions: int
    new_students: int


class SystemOverview(BaseModel):
    total_students: int
    total_teachers: int
    total_learning_sessions: int
    total_grammar_sessions: int
    total_learning_answers: int
    total_grammar_answers: int
    total_words: int
    total_grammar_questions: int
    daily_growth: list[DailyGrowth]


class WordCalibrationItem(BaseModel):
    word_id: str
    english: str
    korean: str
    book_name: str
    lesson: str
    curriculum_level: int
    actual_accuracy: float
    attempt_count: int
    avg_time_sec: Optional[float] = None
    suggested_level: int
    gap: int


class GrammarCalibrationItem(BaseModel):
    question_id: str
    question_type: str
    question_type_label: str
    book_title: str
    chapter_title: str
    assigned_difficulty: int
    actual_accuracy: float
    attempt_count: int
    avg_time_sec: Optional[float] = None
    suggested_difficulty: int
    gap: int


class CalibrationResponse(BaseModel):
    word_calibrations: list[WordCalibrationItem]
    grammar_calibrations: list[GrammarCalibrationItem]


class BadQuestionItem(BaseModel):
    question_id: str
    question_type: str
    question_type_label: str
    book_title: str
    chapter_title: str
    difficulty: int
    accuracy: float
    attempt_count: int
    avg_time_sec: Optional[float] = None
    flag_reason: str


class BadWordItem(BaseModel):
    word_id: str
    english: str
    korean: str
    book_name: str
    curriculum_level: int
    accuracy: float
    attempt_count: int
    flag_reason: str


class BadQuestionResponse(BaseModel):
    grammar_issues: list[BadQuestionItem]
    word_issues: list[BadWordItem]


class ConfusedPair(BaseModel):
    correct_answer: str
    wrong_answer: str
    confusion_count: int


class QuestionTypeAccuracy(BaseModel):
    question_type: str
    label: str
    total: int
    correct: int
    accuracy_pct: float
    avg_time_sec: Optional[float] = None


class ErrorPatternResponse(BaseModel):
    confused_word_pairs: list[ConfusedPair]
    word_question_type_breakdown: list[QuestionTypeAccuracy]
    grammar_question_type_breakdown: list[QuestionTypeAccuracy]


class StageCount(BaseModel):
    stage: int
    count: int


class StageDuration(BaseModel):
    stage: int
    avg_days: float


class SrsOptimizationData(BaseModel):
    stage_distribution: list[StageCount]
    avg_days_per_stage: list[StageDuration]
    total_mastered: int
